README.md
PI PRESENTS  - Version 1.3.5
============================

Diese Readme-Datei hat Peter Vasen ins Deutsche übersetzt. Klicken Sie hier 
http://www.web-echo.de/4.html

This repository contains the stable version Pi Presents. The version  will not be developed further other than bug and compatibility fixes.


TO INSTALL PIPRESENTS-GAPLESS
=============================

Read the 'Installling Pi Presents Gapless' section below.

TO UPGRADE FROM VERSION 1.2 [PIPRESENTS-NEXT]
=============================================
Read the file upgrade_1p2_1p3.txt


TO UPGRADE FROM EARLIER VERSIONS OF PIPRESENTS-GAPLESS
======================================================
To upgrade follow the instructions in the 'Updating Pi Presents' section below. Then follow any instructions in the Release Notes.


PI PRESENTS
===========

Pi Presents is a toolkit for producing interactive multimedia applications for museums, visitor centres, and more.

There are a number of Digital Signage solutions for the Raspberry Pi which are generally browser based, limited to slideshows, non-interactive, and driven from a central server enabling the content to be modified frequently.

Pi Presents is different, it is stand alone, multi-media, highly interactive, diverse in it set of control paradigms – slideshow, cursor controlled menu, radio button, and hyperlinked show, and able to interface with users or machines over several types of interface. It is aimed primarly at curated applications in museums, science centres, and visitor centres.

Being so flexible Pi Presents needs to be configured for your application. This is achieved using a simple to use graphical editor and needs no Python programming. There are numerous tutorial examples and a comprehensive manual.

There are two versions of Pi Presents. ‘Gapless’ is the current stable version. ‘Next’ is now 3 years old and missing many of the refinements and later developments. This will be supported until June 2019.

For a detailed list of applications and features see here:

          https://pipresents.wordpress.com/features/



Licence
=======

See the licence.md file. Pi Presents is Careware to help support a small museum charity of which I am a Trustee http://www.museumoftechnology.org.uk  Particularly if you are using Pi Presents in a profit making situation a donation would be appreciated.


Installing Pi Presents Gapless
===============================

The full manual in English is here https://github.com/KenT2/pipresents-gapless/blob/master/manual.pdf. It will be downloaded with Pi Presents.

There is a German version of the manual written by Peter Vasen ( http://www.web-echo.de/ ) you can download it here

http://www.web-echo.de/4.html


Requirements
-------------

	* must use the latest version of Raspbian Stretch or Buster with Desktop (not the Lite version)
	* must be run from the PIXEL desktop.
	* must be installed and run from user Pi
	
	NOTE: PI PRESENTS DOES NOT WORK WITH BULLSEYE WHICH IS THE LATEST VERSION OF THE OS. INSTEAD USE THE LEGACY VERSION BASED ON BUSTER THAT CAN BE OBTAINED USING THE IMAGER OR CAN BE DOWNLOADED FROM HERE:

https://www.raspberrypi.com/software/operating-systems/#raspberry-pi-os-legacy


Set the GPU Memory size to 256MB
---------------------------------
Using the Raspbian menu preferences>raspberry pi configuration>performance, increase the GPU Memory to 256.


Ensure Raspbian is up to date.
-------------------------------
Pi Presents MUST have the latest version of omxplayer and of Raspbian, get this by

		sudo apt-get update
		sudo apt-get upgrade

Install required packages 
-----------------------------
         sudo apt-get install python-pil (may already be installed)
         sudo apt-get install python-pil.imagetk
         sudo apt-get install unclutter
         sudo apt-get install mplayer
         sudo apt-get install uzbl
         sudo apt-get install python-pexpect (may already be installed)
		 
Install optional packages
------------------------------
         sudo pip install evdev  (if you are using the input device I/O plugin)

	   
Download Pi Presents Gapless
----------------------------

From a terminal window open in your home directory type:

         wget https://github.com/KenT2/pipresents-gapless/tarball/master -O - | tar xz     # -O is a capital Ohhh...

There should now be a directory 'KenT2-pipresents-gapless-xxxx' in your /home/pi directory. Copy or rename the directory to pipresents

Run Pi Presents to check the installation is successful. From a terminal window opened in the home directory type:

         python /home/pi/pipresents/pipresents.py

You will see a window with an error message which is because you have no profiles.Click OK to exit Pi Presents.


Download and try an Example Profile
-----------------------------------

Examples are in the github repository pipresents-gapless-examples.

Open a terminal window in your home directory and type:

         wget https://github.com/KenT2/pipresents-gapless-examples/tarball/master -O - | tar xz

There should now be a directory 'KenT2-pipresents-gapless-examples-xxxx' in the /home/pi directory. Open the directory and move the 'pp_home' directory and its contents to the /home/pi directory.

From the terminal window type:

         python /home/pi/pipresents/pipresents.py -p pp_mediashow_1p3
		 
to see a repeating multimedia show

Exit this with CTRL-BREAK or closing the window, then:

          python /home/pi/pipresents/pipresents.py -p pp_mediashow_1p3 -f -b
		  
to display full screen and to disable screen blanking


Now read the manual to try other examples.


Updating Pi Presents from earlier Versions of Pi Presents Gapless
=================================================================

Open a terminal window in the /home/pi and type:

         wget https://github.com/KenT2/pipresents-gapless/tarball/master -O - | tar xz

There should now be a directory 'KenT2-pipresents-gapless-xxxx' in the /home/pi directory

Rename the existing pipresents directory to old-pipresents

Rename the new directory to pipresents.

Copy pp_editor.cfg, pp_web.cfg, pp_email.cfg, pp_oscmonitor.cfg, and pp_oscremote.cfg from the old to new /pipresents/pp_config directory.


Getting examples for this version.
----------------------------------

Examples are in the github repository pipresents-gapless-examples.

Rename the existing pp_home directory to old_pp_home.

Open a terminal window in your home directory and type:

         wget https://github.com/KenT2/pipresents-gapless-examples/tarball/master -O - | tar xz

There should now be a directory 'KenT2-pipresents-gapless-examples-xxxx' in the /home/pi directory.

Open the directory and move the 'pp_home' directory and its contents to the /home/pi directory.

These examples are compatible with the version of Pi Presents you have just downloaded. In addition you can update profiles from earlier 1.3.x versions by simply opening them in the editor (make a backup copy first).

You can use the update>update all menu option of the editor to update all profiles in a single directory at once.

Lastly you will need to do some manual updating of some of the field values as specified in  ReleaseNotes.txt. Start at the paragraph in ReleaseNotes.txt that introduces your previous version and work forwards



Bug Reports and Feature Requests
================================
I am keen to develop Pi Presents further and would welcome bug reports and ideas for additional features and uses.

Please use the Issues tab on Github https://github.com/KenT2/pipresents-gapless/issues.

For more information on how Pi Presents is being used, Hints and Tips on how to use it and all the latest news hop over to the Pi Presents website https://pipresents.wordpress.com/



changelog.txt
CHANGES FOR 1.3.5i 8/1/2022
============================
fixed  - bug in pp_hyperlinkshow.py when using track timeout
fixed - pp_hyperlinkshow_1p3 example show timeout and track timeout values incorrect.
readme - updated to use Raspberry Pi OS (Legacy)


CHANGES FOR 1.3.5h 2/1/2021
============================
fixed - audio track audio device selection now works with the Raspbian release of DEC 2020, in addition to earlier releases.(see release notes)
fixed - the previous version did not set the audio device from the profile.


CHANGES FOR 1.3.5g 16/6/2020
============================
fixed - audio track audio device selection now works with the Raspbian release of May 2020, in addition to earlier releases.(see release notes)


CHANGES FOR 1.3.5f 1/4/2020
==========================
fixed - In pp_manager.py from the profile>manage or media>manage menus
 in the Manage Dialog. If DELETE is pressed with no file selected and then confirmed
 then all profiles or media are deleted.

CHANGES FOR 1.3.5e 22/9/2019
==========================
tested - Gapless works on Buster with Pi3 and Pi4
fixed - crash in pp_serialdriver.py I/O plugin
fixed - videoplayer volume control initialization sets volume to -60 dB
fixed - crash if format of controls for video track cause an error
fixed - osc commands are ignored if Pi not connected to LAN instead of reporting error
fixed - removed debug messages from various places
fixed - pp_menu_1p4 example, loss of cursor after message track (profile edit)
fixed - adjusted timing for freeze before start and freeze after start in pp_omxdriver.py

CHANGES FOR 1.3.5d 14/4/2019
==========================
fixed - trailing comments delimited by ; not # i keys.cfg, also in example profiles
fixed - crash when terminating Pi Presents from the Manager
fixed - crash when exitpipresents is a Show Command in a show.
fixed - errors in and improvements to the manual

CHANGES FOR 1.3.5c 7/6/2018
============================
modified - readme.md ready for gapless to become the stable version
modified - added upgrade_1p2_1p3.txt containing instructions for upgrade from pp-next to pp-gapless.
enhanced - web editor, dropdown field values displayed in red if read value is not valid
fixed - crash if linked pins are used in gpio.cfg
fixed - crash if I/O plugin .cfg file has errors
fixed - crash if track plugin is not found.

CHANGES FOR 1.3.5b 31/5/2018
============================
enhanced - I/O plugin for PN532 RFID tag reader
enhanced - cec show control commands to turn TV on/standby
fixed - crash if there are no anonymous tracks on a medialst.
fixed - network detection more resilient to peculiar ip route response


CHANGES FOR 1.3.5a 1/05/2018
============================
enhanced - use web editor instead of the schedule.json file to edit a schedule
enhanced - schedule can now execute all show commands
enhanced  - schedule can contain multiple day statements

CHANGES FOR 1.3.4d 25/03/2018
============================
fixed - memory leak when using subshows issue #20
fixed - minor improvements to some examples

CHANGES FOR 1.3.4c 10/03/2018
============================
enhanced - pp_kbddriver_plus enhances pp_kbddriver by allowing strings to trigger events and to be used by track_plugins
enhanced - pp_kbddisplay_1p3 example using pp_kbddriver_plus
enhanced  - krt_camera track plugin allows the Pi'c amera to be used to display real time images and capture pictures
enhanced - pp_camera_1p3 example
modified - debug_autostart.sh now has example text for debugging PP at autostart in addition to pp_manager.py
fixed - pp_serialdrive I/O plugin, individual events can now be disabled
fixed - various quality improvements

CHANGES FOR 1.3.4b 21/02/2018
============================
enhanced - values input by I/O plugins are now available to track plugins via get_input() method. See Release Notes.
enhanced - new serial port I/O plugin
enhanced - new serial port examples pp_serialcontrol_1p3 and pp_serialdisplay_1p3
enhanced  - new I2C I/O plugin supporting ADC, DAC, Pimoroni Four Letter PHat and ScrollHD PHat.
enhanced - new examples for I2C driver pp_i2cioplugin_1p3
fixed - animation commands now work if there is more than one space between fields
fixed - pp_kbddriver now can be disabled
fixed - PP does not crash if time of day scheduler is used with no Start Shows


CHANGES FOR 1.3.4a 13/01/2018
============================
enhanced - pause now has a timeout, set in the profile
enhanced - added closeall, openexclusive and reboot show/OSC commands

CHANGES FOR 1.3.3c 8/12/2017
============================
enhanced - OSC
enhanced - added pp_osc_1p3 example

CHANGES FOR 1.3.3b  21/10/2017
====================
enhanced - documented I/O plugin API in manual and added pp_exampledriver.py demonstrated in pp_ioplugin_1p3 example
enhanced -  animate commands can now have strings containing spaces provided they are surrounded by ""
enhanced - added empty pp_io_config directory to templates
fixed - increase timeout on unpause for show for slow to load video tracks (.avi seems much slower than .mp4)
fixed - crash if an input event happens immeadiatly after PP starts.


CHANGES FOR 1.3.3a  27/09/2017
====================
enhanced - io_plugins. Users can write their own I/O device driver 
enhanced - input_device.py utility assists configuration of devices in /dev/input that use evdev
enhanced  - new i/o plugin (pp_inputdevicedriver.py) for devices in /dev/input that use evdev
enhanced - new .cfg file which configures pp_inputdevicedriver.py specifically for OSMC official remote control.
enhanced - Pi Presents - New method of programming the shutdown pin in gpio.cfg. Allows other device drivers to do shutdown
enhanced - videoplayer, more real time controls added
enhanced - liveshow, added 'escape' track to allow use of stop command when the 'empty' track is a show.
enhanced - Pi Presents - detect there is no user and object

pp_gpio.py
import time
import copy
import os
import ConfigParser
from pp_utils import Monitor
from pp_options import command_options

class PPIO(object):
    """
    PPIO provides some IO facilties for Pi presents
     - configures GPIO pins from data in gpio.cfg
     - reads and debounces inputs pins, provides callbacks on state changes which are used to trigger mediashows
     - for output pins allows players to put events, which request the change of state of pins, into a queue. Events are executed at the required time.
    """
 
 
# constants for buttons

# cofiguration from gpio.cfg
    PIN=0                # pin on RPi board GPIO connector e.g. P1-11
    DIRECTION = 1 # IN/OUT/NONE (None is not used)
    NAME = 2      # name for output
    RISING_NAME=3             # name for rising edge callback
    FALLING_NAME=4      # name ofr falling edge callback
    ONE_NAME=5     # name for one state callback
    ZERO_NAME = 6   # name for zero state callback
    REPEAT =  7   # reperat interval for state callbacks (mS)
    THRESHOLD = 8       # threshold of debounce count for state change to be considered
    PULL = 9                  # pull up or down or none
    
# dynamic data
    COUNT=10          # variable - count of the number of times the input has been 0 (limited to threshold)
    PRESSED = 11      # variable - debounced state 
    LAST = 12       # varible - last state - used to detect edge
    REPEAT_COUNT = 13

    
    TEMPLATE = ['',   # pin
                '',              # direction
                '',              # name
                '','','','',       #input names
                0,             # repeat
                0,             # threshold
                '',             #pull
                0,False,False,0]   #dynamics
    
    PINLIST = ('P1-03','P1-05','P1-07','P1-08',
               'P1-10','P1-11','P1-12','P1-13','P1-15','P1-16','P1-18','P1-19',
               'P1-21','P1-22','P1-23','P1-24','P1-26')

    # index of shutdown pin
    SHUTDOWN_INDEX=0
             
    # constants for sequencer           
    
    SEQUENCER_PIN = 0         # GPIO pin number, the xx in P1-xx
    SEQUENCER_TO_STATE = 1    # False = off , True =on
    SEQUENCER_TIME = 2        # time since the epoch in seconds
    SEQUENCER_TAG = 3   # tag used to delete all matching event, usually a track reference.

# CLASS VARIABLES
    events=[]
    pins=[]
    last_poll_time=0
    options=None
    # gpio_enabled=False

    
    EVENT_TEMPLATE=[0,False,0,None]

    # executed by main program and by each object using gpio
    def __init__(self):
        self.mon=Monitor()
        self.mon.on()
        self.options=command_options()

     # executed once from main program   
    def init(self,pp_dir,pp_home,pp_profile,widget,button_tick,callback=None):
        
        # instantiate arguments
        self.widget=widget
        self.pp_dir=pp_dir
        self.pp_profile=pp_profile
        self.pp_home=pp_home
        self.button_tick=button_tick
        self.callback=callback

        PPIO.SHUTDOWN_INDEX=0

        # read gpio.cfg file.
        if self.read(self.pp_dir,self.pp_home,self.pp_profile) is False:
            return False

        import RPi.GPIO as GPIO
        self.GPIO = GPIO
        
        # construct the GPIO control list from the configuration
        for index, pin_def in enumerate(PPIO.PINLIST):
            pin=copy.deepcopy(PPIO.TEMPLATE)
            pin_bits = pin_def.split('-')
            pin_num=pin_bits[1:]
            pin[PPIO.PIN]=int(pin_num[0])
            if self.config.has_section(pin_def) is False:
                self.mon.log(self, "no pin definition for "+ pin_def)
                pin[PPIO.DIRECTION]='None'            
            else:
                # unused pin
                if self.config.get(pin_def,'direction') == 'none':
                    pin[PPIO.DIRECTION]='none'
                else:
                    pin[PPIO.DIRECTION]=self.config.get(pin_def,'direction')
                    if pin[PPIO.DIRECTION] == 'in':
                        # input pin
                        pin[PPIO.RISING_NAME]=self.config.get(pin_def,'rising-name')
                        pin[PPIO.FALLING_NAME]=self.config.get(pin_def,'falling-name')
                        pin[PPIO.ONE_NAME]=self.config.get(pin_def,'one-name')
                        pin[PPIO.ZERO_NAME]=self.config.get(pin_def,'zero-name')
                        if pin[PPIO.FALLING_NAME] == 'pp-shutdown':
                            PPIO.SHUTDOWN_INDEX=index
                        if self.config.get(pin_def,'repeat') != '':
                            pin[PPIO.REPEAT]=int(self.config.get(pin_def,'repeat'))
                        else:
                            pin[PPIO.REPEAT]=-1
                        pin[PPIO.THRESHOLD]=int(self.config.get(pin_def,'threshold'))
                        if self.config.get(pin_def,'pull-up-down') == 'up':
                            pin[PPIO.PULL]=GPIO.PUD_UP
                        elif self.config.get(pin_def,'pull-up-down') == 'down':
                            pin[PPIO.PULL]=GPIO.PUD_DOWN
                        else:
                            pin[PPIO.PULL]=GPIO.PUD_OFF
                    else:
                        # output pin
                        pin[PPIO.NAME]=self.config.get(pin_def,'name')
 
            # print pin            
            PPIO.pins.append(copy.deepcopy(pin))

        # setup GPIO
        self.GPIO.setwarnings(False)        
        self.GPIO.setmode(self.GPIO.BOARD)
        

        # set up the GPIO inputs and outputs
        for index, pin in enumerate(PPIO.pins):
            num = pin[PPIO.PIN]
            if pin[PPIO.DIRECTION] == 'in':
                self.GPIO.setup(num,self.GPIO.IN,pull_up_down=pin[PPIO.PULL])
            elif  pin[PPIO.DIRECTION] == 'out':
                self.GPIO.setup(num,self.GPIO.OUT)
                self.GPIO.setup(num,False)
        self.reset_inputs()
        PPIO.gpio_enabled=True

        # init timer
        self.button_tick_timer=None
        PPIO.last_scheduler_time=long(time.time())
        return True

    # called by main program only         
    def poll(self):
        # look at the buttons
        self.do_buttons()

        # kick off output pin sequencer
        poll_time=long(time.time())
        # is current time greater than last time the sceduler was run (previous second or more)
        # run in a loop to catch up because root.after can get behind when images are being rendered etc.
        while PPIO.last_scheduler_time<=poll_time:
            self.do_sequencer(PPIO.last_scheduler_time)
            PPIO.last_scheduler_time +=1
        
        # and loop
        self.button_tick_timer=self.widget.after(self.button_tick,self.poll)


    # called by main program only                
    def terminate(self):
        if self.button_tick_timer is not None:
            self.widget.after_cancel(self.button_tick_timer)
        self.clear_events_list(None)
        self.reset_outputs()
        self.GPIO.cleanup()


# ************************************************
# gpio input functions
# called by main program only
# ************************************************
    
    def reset_inputs(self):
        for pin in PPIO.pins:
            pin[PPIO.COUNT]=0
            pin[PPIO.PRESSED]=False
            pin[PPIO.LAST]=False
            pin[PPIO.REPEAT_COUNT]=pin[PPIO.REPEAT]

    # index is of the pins array, provided by the callback ***** needs to be name
    def shutdown_pressed(self):
        if PPIO.SHUTDOWN_INDEX != 0:
            return PPIO.pins[PPIO.SHUTDOWN_INDEX][PPIO.PRESSED]
        else:
            return False

    def do_buttons(self):
        for index, pin in enumerate(PPIO.pins):
            if pin[PPIO.DIRECTION] == 'in':
                # debounce
                if self.GPIO.input(pin[PPIO.PIN]) == 0:
                    if pin[PPIO.COUNT]<pin[PPIO.THRESHOLD]:
                        pin[PPIO.COUNT]+=1
                        if pin[PPIO.COUNT] == pin[PPIO.THRESHOLD]:
                            pin[PPIO.PRESSED]=True
                else: # input us 1
                    if pin[PPIO.COUNT]>0:
                        pin[PPIO.COUNT]-=1
                        if pin[PPIO.COUNT] == 0:
                            pin[PPIO.PRESSED]=False
     
                # detect edges
                # falling edge
                if pin[PPIO.PRESSED] is True and pin[PPIO.LAST] is False:
                    pin[PPIO.LAST]=pin[PPIO.PRESSED]
                    pin[PPIO.REPEAT_COUNT]=pin[PPIO.REPEAT]
                    if  pin[PPIO.FALLING_NAME] != '' and self.callback  is not  None:
                        self.callback(index, pin[PPIO.FALLING_NAME],"falling")
               # rising edge
                if pin[PPIO.PRESSED] is False and pin[PPIO.LAST] is True:
                    pin[PPIO.LAST]=pin[PPIO.PRESSED]
                    pin[PPIO.REPEAT_COUNT]=pin[PPIO.REPEAT]
                    if  pin[PPIO.RISING_NAME] != '' and self.callback  is not  None:
                        self.callback(index, pin[PPIO.RISING_NAME],"rising")

                # do state callbacks
                if pin[PPIO.REPEAT_COUNT] == 0:
                    if pin[PPIO.ZERO_NAME] != '' and pin[PPIO.PRESSED] is True and self.callback is not None:
                        self.callback(index, pin[PPIO.ZERO_NAME],"zero")
                    if pin[PPIO.ONE_NAME] != '' and pin[PPIO.PRESSED] is False and self.callback is not None:
                        self.callback(index, pin[PPIO.ONE_NAME],"zero")
                    pin[PPIO.REPEAT_COUNT]=pin[PPIO.REPEAT]
                else:
                    if pin[PPIO.REPEAT] != -1:
                        pin[PPIO.REPEAT_COUNT]-=1

                    
# ************************************************
# gpio output sequencer functions
# ************************************************

    # execute events at the appropriate time and remove from list (runs from main program only)
    # runs through list a number of times because of problems with pop messing up list
    def do_sequencer(self,schedule_time):
        # print 'sequencer run for: ' + str(schedule_time) + ' at ' + str(long(time.time()))
        while True:
            event_found=False
            for index, item in enumerate(PPIO.events):
                if item[PPIO.SEQUENCER_TIME]<=schedule_time:
                    event=PPIO.events.pop(index)
                    event_found=True
                    self.do_event(event[PPIO.SEQUENCER_PIN],event[PPIO.SEQUENCER_TO_STATE],item[PPIO.SEQUENCER_TIME])
                    break
            if event_found is False: break

    # execute an event
    def do_event(self,pin,to_state,req_time):
        self.mon.log (self,'pin P1-'+ str(pin)+ ' set  '+ str(to_state) + ' required: ' + str(req_time)+ ' actual: ' + str(long(time.time())))
        # print 'pin P1-'+ str(pin)+ ' set  '+ str(to_state) + ' required: ' + str(req_time)+ ' actual: ' + str(long(time.time()))
        self.GPIO.output(pin,to_state)

# ************************************************
# gpio output sequencer interface methods
# these can be called from many classes so need to operate on class variables
# ************************************************
    def animate(self,text,tag):
        if self.options['gpio'] is True:
            lines = text.split("\n")
            for line in lines:
                error_text=self.parse_animate_fields(line,tag)
                if error_text  != '':
                    return 'error',error_text
            return 'normal',''
        return 'normal',''

    # clear event list
    def clear_events_list(self,tag):
        if self.options['gpio'] is True:
            self.mon.log(self,'clear events list ')
            # empty event list
            if tag is None:
                PPIO.events=[]
            else:
                self.remove_events(tag)

    def reset_outputs(self):
        if self.options['gpio'] is True:
            self.mon.log(self,'reset outputs')
            for index, pin in enumerate(PPIO.pins):
                num = pin[PPIO.PIN]
                if pin[PPIO.DIRECTION] == 'out':
                    self.GPIO.output(num,False)

# ************************************************
# internal functions
# these can be called from many classes so need to operate on class variables
# ************************************************

    def parse_animate_fields(self,line,tag):
        fields= line.split()
        if len(fields) == 0:
            return ''
            
        name=fields[0]
        pin= self.pin_of(name)
        if pin  == -1:
            return 'Unknown gpio logical output in: ' + line
       
        to_state_text=fields[1]
        if not (to_state_text  in ('on','off')):
            return 'Illegal to-state in : '+ line
        
        if to_state_text == 'on':
            to_state=True
        else:
            to_state=False
            
        if len(fields) == 2:
            delay_text='0'
        else:
            delay_text=fields[2]
        
        if  not delay_text.isdigit():
            return 'Delay is not an integer in : '+ line
        delay=int(delay_text)
        
        self.add_event(pin,to_state,delay,tag)
        # self.print_events()
        return ''

    def pin_of(self,name):
        for pin in PPIO.pins:
            # print " in list" + pin[PPIO.NAME] + str(pin[PPIO.PIN] )
            if pin[PPIO.NAME] == name and pin[PPIO.DIRECTION] == 'out':
                return pin[PPIO.PIN]
        return -1

    def print_events(self):
        print
        print 'events list'
        for i in PPIO.events:
            print i

    def add_event(self,sequencer_pin,sequencer_to_state,sequencer_time,sequencer_tag):
        poll_time=long(time.time())
        # delay is 0 so just do it, don't queue it.
##        if sequencer_time == 0:
##            print "firing now",poll_time
##            self.do_event(sequencer_pin,sequencer_to_state,poll_time)
##            return
        # prepare the event
        event=PPIO.EVENT_TEMPLATE
        event[PPIO.SEQUENCER_PIN]=sequencer_pin
        event[PPIO.SEQUENCER_TO_STATE]=sequencer_to_state
        event[PPIO.SEQUENCER_TIME]=sequencer_time+poll_time+1
        event[PPIO.SEQUENCER_TAG]=sequencer_tag
        # print 'add event ',event
        # find the place in the events list and insert
        # first item in the list is earliest, if two have the same time then last to be added is fired last.
        abs_time=sequencer_time+poll_time
        copy_event= copy.deepcopy(event)
        for index, item in enumerate(PPIO.events):
            if abs_time<item[PPIO.SEQUENCER_TIME]:
                PPIO.events.insert(index,copy_event)
                return copy_event
        PPIO.events.append(copy_event)
        return copy_event
    
    # remove an event not used and does not work
    def remove_event(self,event):
        for index, item in enumerate(PPIO.events):
            if event == item:
                del PPIO.events[index]
                return True
        return False


    # remove all the events with the same tag, usually a track reference
    def remove_events(self,tag):
        left=[]
        for item in PPIO.events:
            if tag != item[PPIO.SEQUENCER_TAG]:
                left.append(item)
        PPIO.events= left
        # self.print_events()



# ***********************************
# reading gpio.cfg functions
# ************************************

    def read(self,pp_dir,pp_home,pp_profile):
        # try inside profile
        tryfile=pp_profile+os.sep+"gpio.cfg"
        # self.mon.log(self,"Trying gpio.cfg in profile at: "+ tryfile)
        if os.path.exists(tryfile):
            filename=tryfile
        else:
            # try inside pp_home
            # self.mon.log(self,"gpio.cfg not found at "+ tryfile+ " trying pp_home")
            tryfile=pp_home+os.sep+"gpio.cfg"
            if os.path.exists(tryfile):
                filename=tryfile
            else:
                # try inside pipresents
                # self.mon.log(self,"gpio.cfg not found at "+ tryfile + " trying inside pipresents")
                tryfile=pp_dir+os.sep+'pp_home'+os.sep+"gpio.cfg"
                if os.path.exists(tryfile):
                    filename=tryfile
                else:
                    self.mon.log(self,"gpio.cfg not found at "+ tryfile)
                    self.mon.err(self,"gpio.cfg not found")
                    return False   
        self.config = ConfigParser.ConfigParser()
        self.config.read(filename)
        self.mon.log(self,"gpio.cfg read from "+ filename)
        return True


